torch/_library/triton.py

```diff
@@ -130,14 +130,47 @@ def backend_fn(*args, **kwargs):  # type: ignore[no-untyped-def]
         #   can see a call to the triton kernel(s) and so it can directly optimize
         #   them by inlining them into the lowering process.
-        # - With post-dispatch torch.export, this means that there will
-        #   be a call(s) to the triton_kernel_wrapper_functional HOP in the
-        #   graph (that we have yet to figure out how to serialize).
         def functional_decomp(  # type: ignore[no-untyped-def]
-            mode, _, types, args, kwargs
+            mode, op, types, args, kwargs
         ):
-            with mode:
-                return fn(*args, **kwargs)
+            # NOTE [Export custom triton op]
+            # For torch.export (strict and non-strict), we don't do functional decomposition.
+            # Instead, we preserve the custom triton ops as custom ops. This is because we want
+            # the exported program to be high-level and serializable. If we decompose
+            # the custom op to a functional hop and make it a node in exported program,
+            # we need to figure out ways of serializing the hop and its arguments, which can be triton.jited
+            # functions and triton dtypes. This is undesireble because:
+            # - it can be tedious to maintain a layer that serializes the jited function (e.g. with a string) and dtypes.
+            # - exported program will contain the implementation detail (e.g. triton source code) for a specific
+            #   backend (GPU), which is probably at a wrong level of abstraction.
+            # - changes to triton or the serialization logic for triton arguments can be BC breaking
+            #
+            # In the short term, we expect users to have a seperate aot_compile stage that compiles the exported program
+            # into a Cubin file on the same machine that users call export, which does autotuning and removes triton
+            # dependency and serve the model with Cubin. This guarantees that triton changes won't break BC.
+            # In the long term, we may export multiple cubins for the triton op directly.
+
+            from torch.compiler import is_exporting
+
+            if is_exporting():
+                from torch._higher_order_ops.auto_functionalize import (
+                    can_auto_functionalize,
+                    do_auto_functionalize,
+                )
+                from torch._subclasses.functional_tensor import PythonFunctionalizeAPI
+
+                if can_auto_functionalize(op):
+                    return do_auto_functionalize(op, args, kwargs)
+
+                assert (
+                    not op._schema.is_mutable
+                ), "custom triton op need to be auto_functionalized if it's mutable"
+                ctx = PythonFunctionalizeAPI(mode, mode.pre_dispatch)
+                unwrapped_args, unwrapped_kwargs = ctx.unwrap_tensors((args, kwargs))
+                return ctx.wrap_tensors(op(*unwrapped_args, **unwrapped_kwargs))
+            else:
```

Comment:
if is_exporting():
    return mode.__torch_dispatch__(op, types, args ,kwargs)

Reply:
udpated, this works out of box.